OSDOCS-9565: configuring AWS Outposts postinstallation (#71263)
openshift/openshift-docs: modules/create-user-workloads-aws-edge.adoc

// Module included in the following assemblies:
//
// * post_installation_configuration/configuring-aws-outposts.adoc

//To-do: reintegrate installation-extend-edge-nodes-aws-local-zones.adoc with create-user-workloads-aws-edge.adoc. Requires global repo update of any xrefs/includes.

:_mod-docs-content-type: PROCEDURE
[id="create-user-workloads-aws-edge_{context}"]
= Creating user workloads in an Outpost

After you extend an {product-title} in an AWS VPC cluster into an Outpost, you can use edge compute nodes with the label `node-role.kubernetes.io/outposts` to create user workloads in the Outpost.

.Prerequisites

* You have extended an AWS VPC cluster into an Outpost.

* You have access to the cluster using an account with `cluster-admin` permissions.

* You have installed the {oc-first}.

* You have created a compute machine set that deploys edge compute machines compatible with the Outpost environment.

.Procedure

> . Configure a `deployment` resource file named `<application_deployment>.yaml` for an application that you want to deploy to the edge compute node in the edge subnet.

Reviewer: I'd suggest mentioning a specific file name (as opposed to the `<substitute_this_word>` formatting) or removing the "named..." part of this step:

```suggestion
. Configure a `deployment` resource file for an application that you want to deploy to the edge compute node in the edge subnet.
```

> . Configure a `deployment` resource file named `<application_deployment>.yaml` for an application that you want to deploy to the edge compute node in the edge subnet.

Reviewer: Should this be `Deployment`? (applies to all specific API references)

+

> .Example `deployment` resource

Reviewer: ```suggestion
.Example `Deployment` manifest
```
?

        node-role.kubernetes.io/outpost: ''
      tolerations: # <6>
      - key: "node-role.kubernetes.io/outposts"
        operator: "Equal"
        value: ""
        effect: "NoSchedule"
      containers:
        - image: openshift/origin-node
          command:
           - "/bin/socat"
          args:
            - TCP4-LISTEN:8080,reuseaddr,fork
            - EXEC:'/bin/bash -c \"printf \\\"HTTP/1.0 200 OK\r\n\r\n\\\"; sed -e \\\"/^\r/q\\\"\"'
          imagePullPolicy: Always
          name: <application_name>
          ports:
            - containerPort: 8080
          volumeMounts:
            - mountPath: "/mnt/storage"
              name: data
      volumes:
      - name: data
        persistentVolumeClaim:
          claimName: <application_name>
----
<1> Specify a name for your application.
<2> Specify a namespace for your application. The application namespace can be the same as the application name.
<3> Specify the storage class name. For an edge compute configuration, you must use the `gp2-csi` storage class.
<4> Specify a label to identify workloads deployed in the Outpost.
<5> Specify the node selector label that targets edge compute nodes.

> <6> Specify tolerations that match the taints in the compute machine set for your edge compute machines. #The taint only has `key` and `effect`. What are `value` and `operator` doing here and how would the user know to set them?#

Author: Open question

> . Create the `deployment` resource CR by running the following command:

Reviewer: ```suggestion
. Create the `Deployment` resource by running the following command:
```

+
[source,terminal]
----
$ oc create -f <application_deployment>.yaml
----

> . Configure a `service` resource file named `<application_service>.yaml` that exposes a pod from a targeted edge compute node to services that run inside your edge network.

Reviewer: Sugggestion:
```suggestion
. Configure a `Service` object that exposes a pod from a targeted edge compute node to services that run inside your edge network.
```

+

> .Example `service` resource

Reviewer: ```suggestion
.Example `Service` manifest
```

[source,yaml]
----
apiVersion: v1
kind: Service # <1>
metadata:
  name:  <application_name>
  namespace: <application_namespace>
spec:
  ports:
    - port: 80
      targetPort: 8080
      protocol: TCP
  type: NodePort
  selector: # <2>
    app: <application_name>
----
<1> Defines the `service` resource.
<2> Specify the label type to apply to managed pods.

. Create the `service` resource CR by running the following command:
+
[source,terminal]
----
$ oc create -f <application_service>.yaml

> ----

Reviewer: See other comments on similar lines

    spec:
      securityContext:
        seccompProfile:
          type: RuntimeDefault
      nodeSelector: # <6>
        machine.openshift.io/zone-group: ${ZONE_GROUP_NAME}
      tolerations: # <7>
      - key: "node-role.kubernetes.io/outposts"
        operator: "Equal"
        value: ""
        effect: "NoSchedule"
      containers:
        - image: openshift/origin-node
          command:
           - "/bin/socat"
          args:
            - TCP4-LISTEN:8080,reuseaddr,fork
            - EXEC:'/bin/bash -c \"printf \\\"HTTP/1.0 200 OK\r\n\r\n\\\"; sed -e \\\"/^\r/q\\\"\"'
          imagePullPolicy: Always
          name: <application_name>
          ports:
            - containerPort: 8080
          volumeMounts:
            - mountPath: "/mnt/storage"
              name: data
      volumes:
      - name: data
        persistentVolumeClaim:
          claimName: <pvc_name>
----
<1> Specify a namespace for your application.
<2> #What do they need to do for the persistent volume claim?#
<3> Specify the storage class name. For an edge compute configuration, you must use the `gp2-csi` storage class.
<4> Specify a name for your application.

> <5> #How does this apply to Outposts?# Defines the group to where a zone belongs. For example, `us-east-1-iah-1`.

Author: Updated with your YAML from testing, and adjusted the callouts. They still need a review once I push that change.

<6> #How does this apply to Outposts?# Targets edge compute nodes that match the specified labels.
<7> Specify tolerations that match the taints in the compute machine set for your edge compute machines. #The taint only has `key` and `effect`. What are `value` and `operator` doing here and how would the user know to set them?#

. Create the `deployment` resource CR by running the following command:
+
[source,terminal]
----
$ oc create -f <application_deployment>.yaml
----

> . Configure a `service` resource file named `<application_service>.yaml` that exposes a pod from a targeted edge compute node to services that run inside your edge network.

Reviewer 2: We are declaring a NodePort service here and when creating Load Balancers. Do you thing we need to move it to LB section? Or say it may be optional depending of the service type they will be used: exposing with service LB, or through a custom ALB ingress.

Author: This makes me wonder if the user workloads topic belongs at the very end (after we go over ALB)
